core/src/main/scala/org/apache/spark/storage/MemoryStore.scala

```diff
@@ -295,9 +296,9 @@ private[spark] class MemoryStore(blockManager: BlockManager, maxMemory: Long)
       // In this case, we should release the memory after we cache the block there.
       // Otherwise, if we return an iterator, we release the memory reserved here
       // later when the task finishes.
-      if (keepUnrolling) {
-        accountingLock.synchronized {
-          val amountToRelease = currentUnrollMemoryForThisThread - previousMemoryReserved
+      accountingLock.synchronized {
```

Comment:
This grabs the lock unconditionally now; why not put it inside the if block?
I'm not as qualified to review this but it seems like you are always unrolling if the storage level includes disk, not just if it includes memory, which aren't the same thing. I think you might also have to explain the call to `blockManager.memoryStore.releasePendingUnrollMemoryForThisThread()` above.